### ISSUE TITLE
Document MBEDTLS_SSL_VERIFY_OPTIONAL not applicable with TLS 1.3 client

### ISSUE THREAD

**Issue by heppth** [Mbed-TLS/mbedtls#7075]
### Summary
In some cases I know that the server is using a self signed certificate which is not in my ca chain. To be able to connect anyway, I set the authmode as follows:
`mbedtls_ssl_conf_authmode(&ctx->conf, MBEDTLS_SSL_VERIFY_OPTIONAL);`
As described in the documentation, the handshake succeeded and I was able to handle the result via `mbedtls_ssl_get_verify_result()`.

### System information

Mbed TLS version: 3.3.0
Operating system and version: ubuntu

### Expected behavior
Even with TLS 1.3 the handshake should succeed if the peer certificate is not in the ca chain and the authmode `MBEDTLS_SSL_VERIFY_OPTIONAL` is set.

### Actual behavior
`mbedtls_ssl_handshake()` returns -2700 (MBEDTLS_ERR_X509_CERT_VERIFY_FAILED). The following code has no effect:
`mbedtls_ssl_conf_authmode(&ctx->conf, MBEDTLS_SSL_VERIFY_OPTIONAL);`

### Steps to reproduce
Connect to any server with TLS 1.3 support and self signed certificate which is not in ca chain.

### Additional information
As a temporary work around I am using this lines of code:
```
if (ignoreCAErrors)
	mbedtls_ssl_conf_max_tls_version(&ctx->conf, MBEDTLS_SSL_VERSION_TLS1_2);
```

I can see the that `ssl_tls13_validate_certificate()` is completely ignoring the configured authmode if it is used as client. 

I would expect TLS 1.3 to behave the same as TLS 1.2. If the behavior is desired, I would expect a note in the documentation of `mbedtls_ssl_conf_authmode()`.

I did not analyzed if `mbedtls_ssl_conf_ca_cb()` would be an alternative.

**Comment by gilles-peskine-arm** [comment]
I'm not personally familiar with that part of the code, but the lack of support [seems deliberate](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.3.0/library/ssl_tls.c#L1259):

> If the client is configured as TLS 1.3 only with optional verify, return
> bad config.

You should be getting `MBEDTLS_ERR_SSL_BAD_CONFIG` from `mbedtls_ssl_setup()`. Can you please confirm the value returned by `mbedtls_ssl_setup()`?

With my security architect hat, requiring server authentication seems sensible: if the client doesn't authenticate the server, then TLS offers zero security. (But why aren't we rejecting `VERIFY_NONE`?) However this argument applies equally to older protocol versions. And it should be documented.

If you know that certain self-signed certificates are acceptable, you should add them to the CA chain. If that's impractical (for example because there are too many to load in memory), then you should indeed use `mbedtls_ssl_conf_ca_cb()` with a callback that both calls `mbedtls_x509_crt_verify()` with your trusted CA list, and checks the self-signed certificates for your custom validity criteria, and returns a success status if either verification passes.

**Comment by heppth** [comment]
Thank you @gilles-peskine-arm for this information.
My client is configured to allow both: TLS 1.2 and TLS 1.3. My client configuration with `MBEDTLS_SSL_VERIFY_OPTIONAL` does not lead to any error code in `mbedtls_ssl_setup()`.

I will now deal with `mbedtls_ssl_conf_ca_cb()` and `mbedtls_x509_crt_verify()`.

**Comment by heppth** [comment]
Now that's weird! When I set `mbedtls_ssl_conf_ca_cb()`, the callback is never called. Instead I get the error 7680 MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED on handshake.

A look into the sources confirms this:
https://github.com/Mbed-TLS/mbedtls/blob/9cb1aa21c41587618dac2b2886d7362633faa869/library/ssl_tls.c#L1803-L1814

As soon as I set the callback, the ca chain is set to NULL. So setting the callback always leads to the error MBEDTLS_ERR_SSL_CA_CHAIN_REQUIRED if the server supports TLS 1.3.

As far as I understand it now, both `mbedtls_ssl_conf_authmode()` and `mbedtls_ssl_conf_ca_cb()` are irrelevant once the server supports TLS 1.3.

**Comment by gilles-peskine-arm** [comment]
The fact that you can't use both `mbedtls_ssl_conf_ca_cb` and `mbedtls_ssl_conf_ca_chain` together is documented. But if `mbedtls_ssl_conf_ca_cb` is ignored for TLS 1.3, that's another bug (it's not a documented limitation, not mentioned in [`docs/architecture/tls13-support.md`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.3.0/docs/architecture/tls13-support.md), and not a tracked issue). @ronald-cron-arm are you aware of any undocumented limitation about using a CA callback for certificate verification in a TLS 1.3 client?

**Comment by ronald-cron-arm** [comment]
> The fact that you can't use both `mbedtls_ssl_conf_ca_cb` and `mbedtls_ssl_conf_ca_chain` together is documented. But if `mbedtls_ssl_conf_ca_cb` is ignored for TLS 1.3, that's another bug (it's not a documented limitation, not mentioned in [`docs/architecture/tls13-support.md`](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.3.0/docs/architecture/tls13-support.md), and not a tracked issue). @ronald-cron-arm are you aware of any undocumented limitation about using a CA callback for certificate verification in a TLS 1.3 client?

I was not aware of any limitation about using a CA callback but looking at the TLS 1.2 and TLS 1.3 code it seems that indeed this is not supported in TLS 1.3. In TLS 1.3 to validate the certificate we use only `mbedtls_x509_crt_verify_with_profile()` which does not care about CA callback. In TLS 1.2 `mbedtls_x509_crt_verify_with_ca_cb()`  is called.

**Comment by ronald-cron-arm** [comment]
The way I see things go forward:
. document that MBEDTLS_SSL_VERIFY_OPTIONAL and MBEDTLS_SSL_VERIFY_OPTIONAL are not supported in TLS 1.3 on client side.
. add support of CA callback in TLS 1.3.
Would that make the job?

**Comment by heppth** [comment]
> The way I see things go forward: . document that MBEDTLS_SSL_VERIFY_OPTIONAL and MBEDTLS_SSL_VERIFY_NONE are not supported in TLS 1.3 on client side. . add support of CA callback in TLS 1.3. Would that make the job?

Yes it would. However, it would make sense to me that `MBEDTLS_SSL_VERIFY_OPTIONAL` is also possible in TLS 1.3 and behaves the same way.

**Comment by ronald-cron-arm** [comment]
> > The way I see things go forward: . document that MBEDTLS_SSL_VERIFY_OPTIONAL and MBEDTLS_SSL_VERIFY_NONE are not supported in TLS 1.3 on client side. . add support of CA callback in TLS 1.3. Would that make the job?
> 
> Yes it would. However, it would make sense to me that `MBEDTLS_SSL_VERIFY_OPTIONAL` is also possible in TLS 1.3 and behaves the same way.

That's a discussion that needs to happen also. Eventually I'd say that we should either support MBEDTLS_SSL_VERIFY_OPTIONAL both in TLS 1.2 and TLS 1.3 or not at all.

**Comment by gilles-peskine-arm** [comment]
I've filed the lack of callback support as a [separate issue](https://github.com/Mbed-TLS/mbedtls/issues/7079).

> > Yes it would. However, it would make sense to me that MBEDTLS_SSL_VERIFY_OPTIONAL is also possible in TLS 1.3 and behaves the same way.
>
> That's a discussion that needs to happen also. Eventually I'd say that we should either support MBEDTLS_SSL_VERIFY_OPTIONAL both in TLS 1.2 and TLS 1.3 or not at all.

My vote is for “not at all” eventually, i.e. a TLS client only allows `MBEDTLS_SSL_VERIFY_REQUIRED`. But that would be a backward incompatible change, so we can do it only in 4.0. I've [filed an issue for that](https://github.com/Mbed-TLS/mbedtls/issues/7080).

In the meantime, we must either document the difference or allow (and test) `OPTIONAL` and `NONE`. My vote is to align 1.3 with 1.2, for the least surprise when migrating 1.2-only applications to support 1.3.

**Comment by mobsense** [comment]
Posted a quick patch on #8509 8509

**Comment by ronald-cron-arm** [comment]
Closing this issue as we have decided eventually to align the behavior of the TLS 1.2 and TLS 1.3 implementations regarding partial or no server authentication. This will be addressed through #9018.

**Comment by fargiolas** [comment]
> Closing this issue as we have decided eventually to align the behavior of the TLS 1.2 and TLS 1.3 implementations regarding partial or no server authentication. This will be addressed through #9018.

Ronald, maybe I'm missing something, but it's not so clear from #9018. The plan is to finally support VERIFY_NONE and VERIFY_OPTIONAL in TLS1.3 or remove it from TLS1.2 to align it to the former?

**Comment by ronald-cron-arm** [comment]
> > Closing this issue as we have decided eventually to align the behavior of the TLS 1.2 and TLS 1.3 implementations regarding partial or no server authentication. This will be addressed through #9018.
> 
> Ronald, maybe I'm missing something, but it's not so clear from #9018. The plan is to finally support VERIFY_NONE and VERIFY_OPTIONAL in TLS1.3 or remove it from TLS1.2 to align it to the former?

With #9018, we will align TLS 1.3 with TLS 1.2 in 3.6 and development. Then later for 4.0 we will probably remove the support for VERIFY_NONE/OPTIONAL in TLS 1.2 and TLS 1.3 client but still with the possibility to achieve the same (but a bit more work) through callbacks.
